dsp_permissions_scripts/models/group.py

```diff
@@ -2,53 +2,108 @@
+class BuiltinGroup(BaseModel):
+    """Represents a DSP-builtin group, in the form 'knora-admin:ProjectAdmin'"""
+
     model_config = ConfigDict(frozen=True)
 
     prefixed_iri: str
 
-    @model_validator(mode="before")
-    @classmethod
-    def _shorten_iri(cls, data: Any) -> Any:
-        if not isinstance(data, dict):
-            return data
-        data["prefixed_iri"] = data["prefixed_iri"].replace(KNORA_ADMIN_ONTO_NAMESPACE, "knora-admin:")
-        return data
+    @model_validator(mode="after")
+    def _check_regex(self) -> Self:
+        valid_group_names = ["SystemAdmin", "Creator", "ProjectAdmin", "ProjectMember", "KnownUser", "UnknownUser"]
+        prefix, name = self.prefixed_iri.split(":")
+        if prefix != "knora-admin" or name not in valid_group_names:
+            raise InvalidGroupError(f"{self.prefixed_iri} is not a valid group IRI")
+        return self
+
+
+class CustomGroup(BaseModel):
+    """Represents a custom group, in the form 'project-shortname:groupname'"""
+
+    model_config = ConfigDict(frozen=True)
+
+    prefixed_iri: str
 
     @model_validator(mode="after")
-    def _check_regex(self) -> Group:
-        if not self.prefixed_iri.startswith("knora-admin:"):
+    def _check_regex(self) -> Self:
+        if not is_valid_prefixed_group_iri(self.prefixed_iri):
             raise InvalidGroupError(f"{self.prefixed_iri} is not a valid group IRI")
+        if self.prefixed_iri.startswith(("knora-admin:", "knora-base:", "knora-api:")):
+            raise InvalidGroupError(f"{self.prefixed_iri} is not a custom group")
         return self
 
 
-def is_prefixed_group_iri(iri: str) -> bool:
+def group_builder(prefixed_iri: str) -> BuiltinGroup | CustomGroup:
+    """
+    Accepts a prefixed IRI, and converts it to a BuiltinGroup or CustomGroup.
+
+    Args:
+        prefixed_iri: string in the form 'knora-admin:ProjectAdmin' or 'project-shortname:groupname'
+
+    Raises:
+        InvalidGroupError: if the input is neither a valid builtin group nor a valid custom group
+    """
+    if prefixed_iri.startswith("knora-admin:"):
+        return BuiltinGroup(prefixed_iri=prefixed_iri)
+    elif re.search(PREFIXED_IRI_REGEX, prefixed_iri):
+        return CustomGroup(prefixed_iri=prefixed_iri)
+    else:
+        raise InvalidGroupError(f"{prefixed_iri} is not a valid group IRI")
+
+
+def _group_discriminator(v: Any) -> str:
+    """
+    This is only for typing purposes,
+    see https://docs.pydantic.dev/latest/concepts/unions/#discriminated-unions-with-callable-discriminator
+    """
+    if isinstance(v, dict):
+        return "builtin" if v["prefixed_iri"].startswith("knora-admin:") else "custom"
+    else:
+        return "builtin" if getattr(v, "prefixed_iri").startswith("knora-admin:") else "custom"
+
+
+Group: TypeAlias = Annotated[
+    Union[
+        Annotated[BuiltinGroup, Tag("builtin")],
+        Annotated[CustomGroup, Tag("custom")],
+    ],
+    Discriminator(_group_discriminator),
+]
```

Comment:
this is horrible, that this is needed. But I suppose it is needed...

Reply:
I would be very open for a simpler approach, but I don't know any.

Reply:
I don't, sadly. I'm not so fond of pydantic anymore, there days. The "under the hood magic" seems to introduce a lot of indirection and unexpected complications like this...